proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,580 @@
+    System.Span<int> b = x; // Ok, equivalent to `System.Span<int> b = x.AsSpan()`
+}
+
+void M2(in Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // An error, no conversion to System.Span<int>
+}
+
+Buffer10<int> GetBuffer() => default;
+
+void M3()
+{
+    System.ReadOnlySpan<int> a = GetBuffer(); // An error, no conversion to System.ReadOnlySpan<int>
+    System.Span<int> b = GetBuffer(); // An error, no conversion to System.Span<int>
+}
+```
+
+Alternatively, we could say that there is an implicit conversion from expression of a fixed-size buffer type to the following types
+(regardless of vaariable/value classification):
+- ```System.Span<T>```
+- ```System.ReadOnlySpan<T>```
+
+But then, converting a readonly variable to ```System.Span<T>``` or converting a value to either type is a ref-safety error.
+
+
+For example:
+``` C#
+void M1(Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // Ok, equivalent to `System.Span<int> b = x.AsSpan()`
+}
+
+void M2(in Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // An error, readonly mismatch
+}
+
+Buffer10<int> GetBuffer() => default;
+
+void M3()
+{
+    System.ReadOnlySpan<int> a = GetBuffer(); // An error, ref-safety
+    System.Span<int> b = GetBuffer(); // An error, ref-safety
+}
+```
+
+With this approach, less context information is needed for conversion classification. This is going to significantly simplify implementation. 
+
+
+### List patterns
+
+[List patterns](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md) will not be supported for instances
+of fixed-size buffer types.
+
+### Definite assignment checking
+
+Regular definite assignment rules are applicable to variables that have a fixed-size buffer type. 
+
+### An alternative to relying on AsSpan/AsReadOnlySpan helpers
+
+If there is a guarantee that the first element in a fixed-size buffer type is aligned at the beginning of the type (no gap), then compiler can use the
+following code to get a ```Span``` value (instead of relying on presence of ```AsSpan``` helper):
+``` C#
+new Span<element type>(Unsafe.AsPointer(ref <the fixed-size buffer writable variable>), <size of the fixed-size buffer>)
+```
+
+And the following code to get a ```ReadOnlySpan``` value (instead of relying on presence of ```AsReadOnlySpan``` helper):
+``` C#
+new ReadOnlySpan<element type>(Unsafe.AsPointer(ref Unsafe.AsRef(in <the fixed-size buffer writable variable>)), <size of the fixed-size buffer>)
+```
+
+In order to reduce IL size at use sites compiler should be able to add two generic reusable helpers into private implementation detail type and
```

Comment:
You can just directly emit `MemoryMarshal.CreateSpan` call inline and pass the address of the buffer and the length to it.

If you are emitting IL, you do not need the `Unsafe.AsPointer` or `Unsafe.AsRef` calls. These calls exist to make the code compile in C#. They are not necessary when emitting IL.